File: backend/LiveSync.Api/DTOs/DocumentDTOs.cs
using System.ComponentModel.DataAnnotations;

namespace LiveSync.Api.DTOs
{
    public class DocumentDto
    {
        public string Id { get; set; } = string.Empty;
        public string Title { get; set; } = string.Empty;
        public string Content { get; set; } = string.Empty;
        public string OwnerId { get; set; } = string.Empty;
        public string? OwnerName { get; set; }
        public string? ShareCode { get; set; }
        public DateTime CreatedAt { get; set; }
        public DateTime UpdatedAt { get; set; }
        public DateTime? LastEditedAt { get; set; }
        public string? LastEditedBy { get; set; }
        public List<SharedDocumentDto> SharedWith { get; set; } = new();
    }

    public class CreateDocumentRequest
    {
        [Required]
        [StringLength(200)]
        public string Title { get; set; } = string.Empty;

        public string Content { get; set; } = string.Empty;
    }

    public class UpdateDocumentRequest
    {
        [StringLength(200)]
        public string? Title { get; set; }

        public string? Content { get; set; }

        public string? LastEditedBy { get; set; }
    }

    public class SharedDocumentDto
    {
        public string Id { get; set; } = string.Empty;
        public string DocumentId { get; set; } = string.Empty;
        public string DocumentTitle { get; set; } = string.Empty;
        public string UserId { get; set; } = string.Empty;
        public string? UserName { get; set; }
        public DateTime SharedAt { get; set; }
        public string AccessLevel { get; set; } = "View";
    }

    public class AddSharedDocumentRequest
    {
        [Required]
        [StringLength(50)]
        public string ShareCode { get; set; } = string.Empty;
    }

    public class GenerateShareCodeRequest
    {
        [Required]
        [StringLength(50)]
        public string ShareCode { get; set; } = string.Empty;
    }

    public class DocumentContentUpdateRequest
    {
        [Required]
        public string Content { get; set; } = string.Empty;

        public string? LastEditedBy { get; set; }
    }
}

File: backend/livesync-api/src/main/java/com/livesync/api/dto/DocumentDtos.java
package com.livesync.api.dto;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Size;

import java.time.Instant;
import java.util.List;

public final class DocumentDtos {
    private DocumentDtos() {
    }

    public record DocumentDto(String id, String title, String content, String ownerId, String ownerName,
                              String shareCode,
                              String defaultAccessLevel, Instant createdAt, Instant updatedAt, Instant lastEditedAt,
                              String lastEditedBy, List<SharedDocumentDto> sharedWith) {
    }

    public record SharedDocumentDto(String id, String documentId, String documentTitle, String userId, String userName,
                                    Instant sharedAt, String accessLevel) {
    }

    public record CreateDocumentRequest(@NotBlank @Size(max = 200) String title, String content) {
    }

    public record UpdateDocumentRequest(@Size(max = 200) String title, String content, String lastEditedBy) {
    }

    public record DocumentContentUpdateRequest(@NotNull String content, String lastEditedBy) {
    }

    public record AddSharedDocumentRequest(@NotBlank @Size(max = 50) String shareCode) {
    }

    public record UpdateAccessLevelRequest(@NotBlank String accessLevel) {
    }

    public record ExecuteDocumentRequest(
            @NotBlank @Pattern(regexp = "^(csharp|cs|python|py|javascript|js|node)$", message = "Supported languages: python, javascript, csharp.") String language,
            @Size(max = 4000) String standardInput) {
    }

    public record ExecutionLanguageDescriptor(String id, String name, String displayName) {
    }

    public record DocumentExecutionResponse(String documentId, String language, String status, boolean isSuccess,
                                            String message, String standardOutput, String standardError,
                                            Double executionDurationMs, Long peakMemoryBytes, Double cpuTimeMs,
                                            String timeComplexity, String spaceComplexity, String complexityExplanation,
                                            Instant requestedAt, Instant completedAt) {
    }

    public record AiAnalysisRequest(String action, String language) {
    }

    public record AiAnalysisResponse(String action, String language, String explanation, List<String> suggestions, String generatedCode) {
    }
}
